### DISCUSSION TITLE
Brainstorming for DB

### DISCUSSION THREAD

**Discussion by BECATRUE** [snu-quiqcl/qiwis#22] [Ideas]
I thought about how to implement `database`.

In my opinion, using `sqlite3` is the best solution!
It is easy to connect to python and since it is just a file, we can use the database easily. (e.g. extract datas, draw a plot using other programs)
However, it's just a file system, so remote access may be difficult. (see [connect to remote sqlite3](https://stackoverflow.com/questions/2318315/connect-to-a-remote-sqlite3-database-with-python))

For more details, please refer to [sqlite docs](https://www.sqlite.org/index.html) and [python docs for sqlite3](https://docs.python.org/ko/3.7/library/sqlite3.html).
And please feel free to leave your opinion!

**Comment by BECATRUE** [comment]
Please check the above suggestion!

@kangz12345

**Reply by kangz12345** [reply]
Okay, for our example cases, local file system looks enough!
However, for example, if both the number generator and the poller write to the same database, we should guarantee the safety.

**Reply by BECATRUE** [reply]
I looked it up and sqlite3 seems to guarantee the safety for concurrency.

[begin concurrent](https://www.sqlite.org/cgi/src/doc/begin-concurrent/doc/begin_concurrent.md)
[sqlite concurrent access](https://stackoverflow.com/questions/4060772/sqlite-concurrent-access)

**Reply by BECATRUE** [reply]
I think we should make a database manager object in backend like [this](https://m.blog.naver.com/tommybee/220730261512).

**Reply by kangz12345** [reply]
> I looked it up and sqlite3 seems to guarantee the safety for concurrency.
> 
> [begin concurrent](https://www.sqlite.org/cgi/src/doc/begin-concurrent/doc/begin_concurrent.md) [sqlite concurrent access](https://stackoverflow.com/questions/4060772/sqlite-concurrent-access)

It really looks so!

**Reply by kangz12345** [reply]
> I think we should make a database manager object in backend like [this](https://m.blog.naver.com/tommybee/220730261512).

Why do we need such a manager? What is the problem when multiple writer threads exist?
Is it a performance problem or correctness problem?

**Reply by BECATRUE** [reply]
Oh.. You are right! We don't need such a manager.

Based on this discussion, I will make an issue.
